pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -380,186 +243,46 @@ func (c *Cluster) getExistingOSDDeploymentsOnPVCs() (map[string]*apps.Deployment
-	ctx := context.TODO()
-	osds := status.OSDs
-	logger.Infof("starting %d osd daemons on pvc %s", len(osds), pvcName)
 	osdProps, err := c.getOSDPropsForPVC(pvcName)
 	if err != nil {
-		config.addError(fmt.Sprintf("%v", err))
-		return
+		return errors.Wrapf(err, "failed to get properties for %s", osdLongName)
 	}
 
-	// start osds
-	for _, osd := range osds {
-		logger.Debugf("start osd %v", osd)
-
-		dp, err := c.makeDeployment(osdProps, osd, config)
-		if err != nil {
-			errMsg := fmt.Sprintf("failed to create deployment for pvc %q. %v", osdProps.crushHostname, err)
-			config.addError(errMsg)
-			continue
-		}
-
-		// Set the deployment hash as an annotation
-		err = patch.DefaultAnnotator.SetLastAppliedAnnotation(dp)
-		if err != nil {
-			errMsg := fmt.Sprintf("failed to set annotation for deployment %q. %v", dp.Name, err)
-			config.addError(errMsg)
-			continue
-		}
-
-		_, createErr := c.context.Clientset.AppsV1().Deployments(c.clusterInfo.Namespace).Create(ctx, dp, metav1.CreateOptions{})
-		if createErr != nil {
-			if kerrors.IsAlreadyExists(createErr) {
-				logger.Infof("deployment for osd %d already exists. updating if needed", osd.ID)
-				if err = updateDeploymentAndWait(c.context, c.clusterInfo, dp, opconfig.OsdType, strconv.Itoa(osd.ID), c.spec.SkipUpgradeChecks, c.spec.ContinueUpgradeAfterChecksEvenIfNotHealthy); err != nil {
-					logger.Errorf("failed to update osd deployment %d. %v", osd.ID, err)
-				}
-			} else {
-				// we failed to create job, update the orchestration status for this pvc
-				logger.Warningf("failed to create osd deployment for pvc %q, osd %v. %v", osdProps.pvc.ClaimName, osd, createErr)
-				continue
-			}
-		}
+	// TODO: why was there not a Progressing condition added here before?
+	// opconfig.ConditionExport(c.context, c.clusterInfo.NamespacedName(), cephv1.ConditionProgressing, v1.ConditionTrue, "ClusterProgressing", fmt.Sprintf("Processing node %s osd %d", nodeName, osd.ID))
 
-		logger.Infof("started deployment for osd %d on pvc", osd.ID)
+	if err := createOrUpdateOSDDeployment(c, osd, osdProps, osdLongName, config); err != nil {
+		return err
 	}
+
+	return nil
 }
 
-func (c *Cluster) startOSDDaemonsOnNode(nodeName string, config *provisionConfig, configMap *v1.ConfigMap, status *OrchestrationStatus) {
-	ctx := context.TODO()
-	osds := status.OSDs
-	logger.Infof("starting %d osd daemons on node %s", len(osds), nodeName)
+func createOrUpdateOSDDaemonOnNode(c *Cluster, osd OSDInfo, nodeName string, config *provisionConfig) error {
+	osdLongName := fmt.Sprintf("OSD %d on node %q", osd.ID, nodeName)
 
 	// fully resolve the storage config and resources for this node
 	n := c.resolveNode(nodeName)
 	if n == nil {
-		logger.Errorf("node %q did not resolve to start osds", nodeName)
-		return
+		// NOTE: this now registers as a failure instead of just printing a warning message
+		return fmt.Errorf("failed to resolve node config to create/update deployment for %s", osdLongName)
 	}
```

Comment:
Thoughts @travisn @leseb ?

Reply:
This will abort the OSD reconcile? Or does it continue with other OSDs?

Reply:
This will **not** abort the reconcile. `createNewOSDsFromStatus` calls this function and adds errors returned to the error collection. This is another low-level method that lets the calling function decide what is a critical error or not.